### PR TITLE
[Numpy] Add op fmax, fmin, fmod

### PR DESCRIPTION
**Pull request by hanke580**
## Description ##
fmax operator

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

Performance comparisons (in microseconds):
Operator | Official NumPy | Legacy ctypes | New ctypes | New Cython
-- | -- | -- | -- | --
fmax |0.52 | 22.91 | 15.89 | 8.13
fmin |0.52 | 22.92 | 15.8 | 8.17
fmod |0.83 | 23.08 | 16.01 | 8.19

### CODE REVIEW

File: src/operator/mshadow_op.h [hunk 1]
```diff
@@ -1107,6 +1107,21 @@ struct maximum : public mxnet_op::tunable {
...
+/*! \brief used for computing binary operator fmax */
+struct fmax : public mxnet_op::tunable {
+  template<typename DType>
+  MSHADOW_XINLINE static DType Map(DType a, DType b) {
+    if (IsNan(b)) {
+      return a;
+    } else if (IsNan(a)) {
+      return b;
+    }
+    else {
```
Comment (haojin2) [inline src/operator/mshadow_op.h]: `} else {`

File: src/operator/numpy/np_elemwise_broadcast_op_extended.cc [hunk 2]
```diff
@@ -371,5 +371,36 @@ MXNET_OPERATOR_REGISTER_BINARY(_backward_npi_rldexp_scalar)
...
+.set_attr<nnvm::FInplaceOption>("FInplaceOption",
+  [](const NodeAttrs& attrs){
+    return std::vector<std::pair<int, int> >{{0, 1}};
+  })
+.set_attr<FResourceRequest>("FResourceRequest",
+  [](const NodeAttrs& attrs) {
+    return std::vector<ResourceRequest>{ResourceRequest::kTempSpace};
+  })
+.set_attr<FCompute>("FCompute<cpu>", BinaryBroadcastBackwardUseIn<cpu, mshadow_op::ge,
+                                                              mshadow_op::lt>);
```
Comment (haojin2) [inline src/operator/numpy/np_elemwise_broadcast_op_extended.cc]:
```c++
.set_attr<FCompute>("FCompute<cpu>", BinaryBroadcastBackwardUseIn<cpu, mshadow_op::ge,
                                                                  mshadow_op::lt>);
```

Comment (hanke580) [reply src/operator/numpy/np_elemwise_broadcast_op_extended.cc]: Resolved Thx

File: src/operator/numpy/np_elemwise_broadcast_op_extended.cc [hunk 3]
```diff
@@ -371,5 +371,71 @@ MXNET_OPERATOR_REGISTER_BINARY(_backward_npi_rldexp_scalar)
...
+.set_attr_parser([](NodeAttrs *attrs) { attrs->parsed = std::stod(attrs->dict["scalar"]); })
+.set_attr<FCompute>("FCompute<cpu>", BinaryScalarOp::Backward<cpu, mshadow_op::ge>);
+
+MXNET_OPERATOR_REGISTER_BINARY_BROADCAST(broadcast_fmin)
+.add_alias("_npi_fmin")
+.set_attr<FCompute>("FCompute<cpu>", BinaryBroadcastCompute<cpu, mshadow_op::fmin>)
+.set_attr<nnvm::FGradient>("FGradient", ElemwiseGradUseIn{"_backward_broadcast_fmin"});
+
+NNVM_REGISTER_OP(_backward_broadcast_fmin)
+.add_alias("_backward_npi_fmin")
```
Comment (haojin2) [inline src/operator/numpy/np_elemwise_broadcast_op_extended.cc]: no need for the alias.

File: src/operator/numpy/np_elemwise_broadcast_op_extended.cc [hunk 4]
```diff
@@ -371,5 +371,71 @@ MXNET_OPERATOR_REGISTER_BINARY(_backward_npi_rldexp_scalar)
 .set_attr_parser([](NodeAttrs *attrs) { attrs->parsed = std::stod(attrs->dict["scalar"]); })
 .set_attr<FCompute>("FCompute<cpu>", BinaryScalarOp::Backward<cpu, mshadow_op::rldexp_grad>);
 
+MXNET_OPERATOR_REGISTER_BINARY_BROADCAST(broadcast_fmax)
```
Comment (haojin2) [inline src/operator/numpy/np_elemwise_broadcast_op_extended.cc]: please register as `_npi_fmax`

File: src/operator/numpy/np_elemwise_broadcast_op_extended.cc [hunk 5]
```diff
@@ -371,5 +371,71 @@ MXNET_OPERATOR_REGISTER_BINARY(_backward_npi_rldexp_scalar)
...
+.add_alias("_npi_fmax")
+.set_attr<FCompute>("FCompute<cpu>", BinaryBroadcastCompute<cpu, mshadow_op::fmax>)
+.set_attr<nnvm::FGradient>("FGradient", ElemwiseGradUseIn{"_backward_broadcast_fmax"});
+
+NNVM_REGISTER_OP(_backward_broadcast_fmax)
```
Comment (haojin2) [inline src/operator/numpy/np_elemwise_broadcast_op_extended.cc]: simply use `_backward_npi_fmax`, and get rid of the alias, similar for other ops.

File: python/mxnet/symbol/numpy/_symbol.py [hunk 6]
```diff
@@ -4156,6 +4171,11 @@ def any(a, axis=None, out=None, keepdims=False):
         in which case a reference to out is returned.
     """
     return _npi.any(a, axis=axis, keepdims=keepdims, out=out)
+=======
```
Comment (haojin2) [inline python/mxnet/symbol/numpy/_symbol.py]: this conflict was not resolved properly.

File: python/mxnet/ndarray/numpy/_op.py [hunk 7]
```diff
@@ -1164,6 +1165,34 @@ def mod(x1, x2, out=None, **kwargs):
...
+
+    Returns
+    -------
+    out : ndarray or scalar
+        This is a scalar if both x1 and x2 are scalars.
+    """
+    if isinstance(x1, numeric_types) and isinstance(x2, numeric_types):
+        _np.fmod(x1, x2, out=out)
+    return _api_internal.fmod(x1, x2, out)
+
```
Comment (haojin2) [inline python/mxnet/ndarray/numpy/_op.py]: one more blank line below.

File: python/mxnet/ndarray/numpy/_op.py [hunk 8]
```diff
@@ -4367,6 +4396,26 @@ def maximum(x1, x2, out=None, **kwargs):
...
+        or shapes that can be broadcast to a single shape.
+
+    Returns
+    -------
+    out : mxnet.numpy.ndarray or scalar
+        The maximum of x1 and x2, element-wise. This is a scalar if both x1 and x2 are scalars."""
+    if isinstance(x1, numeric_types) and isinstance(x2, numeric_types):
+        _np.fmax(x1, x2, out=out)
+    return _api_internal.fmax(x1, x2, out)
+
```
Comment (haojin2) [inline python/mxnet/ndarray/numpy/_op.py]: one more blank line below.

File: python/mxnet/symbol/numpy/_symbol.py (hunk 6 above)
Comment (hanke580) [reply python/mxnet/symbol/numpy/_symbol.py]: Sorry for that, Fixed now.
Thx

**Pull request merged**